src/mlpack/methods/preprocess/preprocess_string_main.cpp

```diff
@@ -0,0 +1,266 @@
+  boost::string_view token, copy;
+  while (std::getline(fin, line))
+  {
+    stringstream streamLine(line);
+    dataset.push_back(std::vector<std::string>());
+    // delimeter[0] becase the standard function accepts char as input.
+    while (std::getline(streamLine, word, column_delimiter[0]))
+    {
+      dataset.back().push_back(word);
+    }
+  }
+  std::set<size_t>::iterator dimensionIterator = dimension.begin();
+  // Preparing the input dataset on which string manipulation has to be done.
+  std::vector<std::vector<std::string>> input(dimension.size());
+  for (size_t i = 0; i < dataset.size(); i++)
+  {
+    size_t col = 0;
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0; j < dataset[i].size(); j++)
+    {
+      if (col < dimension.size())
+      {
+        if (*dimensionIterator == j)
+        {
+          // insert into input
+          input[col].push_back(dataset[i][j]);
+          col++;
+          dimensionIterator++;
+        }
+      }
+      else
+      {
+        break;
+      }
+    }
+  }
+  data::StringCleaning obj;
+  if (CLI::HasParam("lowercase"))
+  {
+    for (size_t i = 0; i < input.size(); i++)
+      obj.LowerCase(input[i]);
+  }
+  if (CLI::HasParam("stopwords"))
+  {
+    // Not sure how to take input for tokenizer from cli.
+    if (!CLI::HasParam("stopwordsfile"))
+    {
+      Log::Fatal << "Please provide a file for stopwords.\n";
+    }
+    ifstream stopwordfile;
+    // Open an existing file
+    const std::string stopwordfilename =
+        CLI::GetParam<std::string>("stopwordsfile");
+    stopwordfile.open(stopwordfilename);
+    if (!stopwordfile.is_open())
+    {
+      Log::Fatal << "Unable to open the file for stopwords.\n";
+    }
+    std::string word;
+    std::deque<std::string> originalword;
+    std::unordered_set<boost::string_view,
+        boost::hash<boost::string_view> >stopwords;
+    while (stopwordfile >> word)
+    {
+      originalword.push_back(word);
+      stopwords.insert(originalword.back());
+    }
+    const std::string delimiter = CLI::GetParam<std::string>
+        ("delimiter");
+    for (size_t i   = 0; i < input.size(); i++)
+      obj.RemoveStopWords(input[i], stopwords,
+          data::SplitByChar(delimiter));
+  }
+  if (CLI::HasParam("punctuation"))
+  {
+    for (size_t i = 0; i < input.size(); i++)
+      obj.RemovePunctuation(input[i]);
+  }
+  const std::string outputFilename = CLI::GetParam<std::string>("preprocess"
+      "_dataset");
+  ofstream fout;
+  fout.open(outputFilename, ios::trunc);
+  if (!fout.is_open())
+  {
+    Log::Fatal << "Unable to open a file for writing output.\n";
+  }
+  dimensionIterator = dimension.begin();
+  for (size_t i = 0 ; i < dataset.size(); i++)
+  {
+    size_t col = 0;
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0 ; j < dataset[i].size(); j++)
+    {
+      if (col < dimension.size() && *dimensionIterator == j)
+      {
+        fout<<input[col][i]<<column_delimiter;
```

Comment:
This will never happen if `dimension` has negative values. And you needn't write the delimiter if this is the last column.
I think this line may lead to segmentation fault since you didn't verify that `input[col]` has `dataset[i].size()` elements.